### PR TITLE
feat(charts): allow query mutator to update queries after splitting original sql

### PR DESCRIPTION
**Pull request by AkashN7**
### SUMMARY

The changes allow a user to use the SQL_QUERY_MUTATOR function to add statements such as SET ROLE without breaking functionality. Currently, adding a statemen like this works fine in the SQL Editor, but Charts run individual statements by splitting a passed query and throw errors since SET ROLE does not return any data on its own. With the new MUTATE_AFTER_SPLIT variable, users can choose to apply the mutator function after the chart's query execution function splits a given query, allowing users to apply a change to all statements in a query and have Chart functionality not be interrupted.

### TESTING INSTRUCTIONS

In the config file, update the SQL_QUERY_MUTATOR function to add a statement which does not return any results on its own and is something which should be run alongside all other statements in a given query. An example of this is adding a SET ROLE statement to the beginning of a query. Create a new Chart and see if queries return errors due to no data being fetched. Now in the config file, add a MUTATE_AFTER_SPLIT variable and set it equal to True. Create a new Chart and make a query, and verify that the desired results are showing.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -916,7 +916,8 @@ def mutate_query_from_config(self, sql: str) -> str:
 
         Typically adds comments to the query with context"""
         sql_query_mutator = config["SQL_QUERY_MUTATOR"]
-        if sql_query_mutator:
+        mutate_after_split = config["MUTATE_AFTER_SPLIT"]
```
Comment (nytai) [inline superset/connectors/sqla/models.py:919]: New config vars should be added to config.py with a default value and a comment explaining what it does.

**Comment by merkinrob** [comment]
@AkashN7  there are conflicts also

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (AkashN7) [reply superset/connectors/sqla/models.py:925]: config.py has been updated with a default value and comment on usage

File: superset/models/core.py [hunk 2]
```diff
@@ -415,7 +415,11 @@ def get_df(  # pylint: disable=too-many-locals
         mutator: Optional[Callable[[pd.DataFrame], None]] = None,
     ) -> pd.DataFrame:
         sqls = self.db_engine_spec.parse_sql(sql)
-        engine = self.get_sqla_engine(schema)
+
+        engine = self.get_sqla_engine(schema=schema, user_name=username)
```
Comment (john-bodley) [inline superset/models/core.py:419]: `user_name` is deprecated in `get_sqla_engine`.

Comment (AkashN7) [reply superset/models/core.py]: I have updated the call to `get_sqla_engine`

**Comment by baz-bakerhughes** [comment]
@john-bodley @nytai How can we get this to the next level our colleague was an intern whose internship has ended and we would like to get this merged into main superset.

File: superset/models/core.py (hunk 2 above)
Comment (baz-bakerhughes) [reply superset/models/core.py]: @john-bodley  @nytai  How can we get this to the next level our colleague was an intern whose internship has ended and we would like to get this merged into main superset. Apologies for my ignorance on what steps are pending on our part if any.  But as none of us are approved reviewers we need a bit of guidance on how to get this into main.

**Review by baz-bakerhughes — APPROVED** [review]

File: superset/models/core.py [hunk 3]
```diff
@@ -415,7 +415,11 @@ def get_df(  # pylint: disable=too-many-locals
         mutator: Optional[Callable[[pd.DataFrame], None]] = None,
     ) -> pd.DataFrame:
         sqls = self.db_engine_spec.parse_sql(sql)
+
         engine = self.get_sqla_engine(schema)
+        username = utils.get_username() or username
```
Comment (villebro) [inline superset/models/core.py:420]: This variable doesn't appear to be defined in this scope?
```suggestion
        username = utils.get_username()
```

File: superset/connectors/sqla/models.py [hunk 4]
```diff
@@ -922,7 +922,9 @@ def mutate_query_from_config(self, sql: str) -> str:
...
+        if sql_query_mutator and not mutate_after_split:
+            username = utils.get_username()
```
Comment (villebro) [inline superset/connectors/sqla/models.py:927]: This variable appears to be redundant as it's not referenced anywhere below
```suggestion
```

File: superset/models/core.py [hunk 5]
```diff
@@ -438,12 +442,29 @@ def _log_query(sql: str) -> None:
...
+                    sqls[-1],
+                    user_name=username,
+                    security_manager=security_manager,
+                    database=None
+                )
+                _log_query(last_sql)
+                self.db_engine_spec.execute(cursor, last_sql)
+            else:
+                _log_query(sqls[-1])
+                self.db_engine_spec.execute(cursor, sqls[-1])
```
Comment (villebro) [inline superset/models/core.py:467]: It appears `sqls[-1]` is needed in both branches of the if statement. This should simplify it somewhat
```suggestion
            last_sql = sqls[-1]
            if mutate_after_split:
                last_sql = sql_query_mutator(
                    last_sql,
                    user_name=username,
                    security_manager=security_manager,
                    database=None
                )
                _log_query(last_sql)
                self.db_engine_spec.execute(cursor, last_sql)
            else:
                _log_query(last_sql)
                self.db_engine_spec.execute(cursor, last_sql)
```

**Review by villebro — CHANGES_REQUESTED** [review]
I left a few comments. I'm sorry for being late to the party, but I can try to help address any remaining issues if the original author is not able to do so.

I think I understand the use case, but unfortunately I wasn't able to reproduce based on the testing instructions. Specifically, I didn't understand how the `SET ROLE` statement was injected into the chart query (AFAIK adding it to the query mutator doesn't get included in the split, i.e. it's only applied after the split). Could we expand the testing instructions with a more detailed explanation of how this has been achieved?

**Comment by merkinrob** [comment]
Thank you @villebro we will get back to you ASAP

**Comment by villebro** [comment]
Closing as a rewrite of this PR has already been merged.

**Comment by baz-bakerhughes** [comment]
Thank you sir.  Yes apologies we should have closed it as this was deprecated and a new one was created because of branch permissions etc.

**Pull request closed without merging**